pkg/injector/injector.go

```diff
@@ -183,42 +187,45 @@ func getServiceAccount(ctx context.Context, kubeClient kubernetes.Interface, all
 	return allowedUids, nil
 }
 
-func (i *injector) Run(ctx context.Context, onReady func()) {
-	go func() {
-		select {
-		case <-ctx.Done():
-			log.Info("Sidecar injector is shutting down")
-			shutdownCtx, cancel := context.WithTimeout(
-				context.Background(),
-				time.Second*5,
-			)
-			defer cancel()
-			err := i.server.Shutdown(shutdownCtx)
-			if err != nil {
-				log.Errorf("Error while shutting down injector: %v", err)
-			}
-		}
-	}()
+func (i *injector) Run(ctx context.Context) error {
+	select {
+	case <-i.ready:
+		return errors.New("injector already running")
+	default:
+		// Nop
+	}
 
 	ln, err := net.Listen("tcp", i.server.Addr)
 	if err != nil {
-		log.Fatalf("Eror while creating listener: %v", err)
+		return fmt.Errorf("error while creating listener: %w", err)
 	}
 
 	log.Infof("Sidecar injector is listening on %s, patching Dapr-enabled pods", i.server.Addr)
 
-	if onReady != nil {
-		onReady()
-	}
-
-	err = i.server.ServeTLS(ln, i.config.TLSCertFile, i.config.TLSKeyFile)
-	if err != http.ErrServerClosed {
-		log.Errorf("Sidecar injector error: %s", err)
-	}
+	errCh := make(chan error, 1)
+	go func() {
+		srverr := i.server.ServeTLS(ln, i.config.TLSCertFile, i.config.TLSKeyFile)
+		if !errors.Is(srverr, http.ErrServerClosed) {
+			errCh <- fmt.Errorf("sidecar injector error: %s", srverr)
+			return
+		}
+		errCh <- nil
+	}()
 
-	ln.Close()
+	close(i.ready)
 
-	log.Info("Sidecar injector stopped")
+	select {
+	case <-ctx.Done():
+		log.Info("Sidecar injector is shutting down")
+		shutdownCtx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
```

Comment:
Consider reducing this timeout. 30s is quite a lot of time for something like the injector IMHO